src/pmpo_c.cpp

```diff
@@ -169,6 +149,62 @@ void polympo_createMPs_f(MPMesh_ptr p_mpmesh,
   p_MPs->setElmIDoffset(offset);
 }
 
+void polympo_rebuildMPs_f(MPMesh_ptr p_mpmesh,
+                        const int numMPs, // total number of MPs which is GREATER than or equal to number of active MPs
+                        const int* allMP2Elm,
+                        const int* addedMPMask) {
+  checkMPMeshValid(p_mpmesh);
+  auto p_MPs = ((polyMPO::MPMesh*)p_mpmesh)->p_MPs;
+  PMT_ALWAYS_ASSERT(numMPs >= p_MPs->getCount());
+  PMT_ALWAYS_ASSERT(numMPs >= p_MPs->getMaxAppID());
+
+  int offset = p_MPs->getElmIDoffset();
+  std::vector<int> added_mpIDs(numMPs);
+  std::vector<int> added_mp2Elm(numMPs);
+  int numAddedMPs = 0;
+  for(int i=0; i<numMPs; i++) {
+    if(addedMPMask[i] == MP_ACTIVE) {
+      added_mpIDs[numAddedMPs] = i;
+      added_mp2Elm[numAddedMPs] = allMP2Elm[i]-offset; //adjust for 1 based indexing if needed
+      numAddedMPs++;
+    }
+  }
+
+  int internalMPCapacity = p_MPs->getCapacity(); // pumipic expects full capacity to rebuild
+  Kokkos::View<int*> mp2Elm("mp2Elm", internalMPCapacity);
+  auto mpAppID = p_MPs->getData<polyMPO::MPF_MP_APP_ID>();
+
+  using space_t = Kokkos::DefaultExecutionSpace::memory_space;
+  auto added_mp2Elm_d = create_mirror_view_and_copy(space_t(), added_mp2Elm.data(), numAddedMPs);
+  auto added_mpIDs_d = create_mirror_view_and_copy(space_t(), added_mpIDs.data(), numAddedMPs);
+  auto addedMPMask_d = create_mirror_view_and_copy(space_t(), addedMPMask, numMPs);
+  auto mpMP2ElmIn_d = create_mirror_view_and_copy(space_t(), allMP2Elm, numMPs);
+
+  auto setMP2Elm = PS_LAMBDA(const int& elm, const int& mp, const int& mask) {
+    if(mask) {
+      if (addedMPMask_d[mpAppID(mp)] == MP_ACTIVE)
+        mp2Elm(mp) = MP_DELETE;
```

Comment:
This logic (setting `MP_DELETE` when an MP is set as `MP_ACTIVE`) isn't obvious at first glance; we may want to add a short comment about what the idea is here for this conditional.

Reply:
I added a comment